Wrapper class file

ninakokalj opened the pull request:
This PR is for the wrapper class, which gets the extractors and anonymization strategy as the input, and has methods for traversing across folders and files and applying the anonymization pipeline on the files.


---

Closes:
#1

ninakokalj committed:
Wrapper class file

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+from .utils.file_system import *
+import json
+from typing import Union, List
+import os
```
Order the imports in the following order: 

1. Python native modules
2. Installed modules via requirements.txt
3. Implemented modules

```python
import os
import json
from typing import Union, List

from .anonymize.extractors import ExtractorInterface, MultiExtractor
from .anonymize.strategies import StrategyInterface
from .utils.file_system import open_file, write_file
```

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+from .anonymize.extractors import ExtractorInterface, MultiExtractor
+from .anonymize.strategies import StrategyInterface
+from .utils.file_system import *
```
Never import all functions using star `*`. It can happen that another module also imports the same-named functions, and you then do not know which is which.

Better to import the actual functions as:

```python
from .utils.file_system import open_file, write_file
```

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+# Pipeline class
+# =====================================
+
+class Pipeline():
```
In Python, no need for brackets if the class does not inherit from another class. Please update to:
```python
class Pipeline:
    ...
```

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+            strategy (StrategyInterface): The strategy to use for anonymization.
+
+        Methods:
+            anonymize(input_dir, output_dir, flatten=False): Anonymize files in the input directory and save the anonymized files to the output directory. 
```
Put the description in new line:

```
anonymize(input_dir, output_dir, flatten=False): 
    Anonymize files in the input directory and save the anonymized files to the output directory. 
```

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+            if isinstance(extractor, ExtractorInterface):
+                self.extractor = extractor
+            elif isinstance(extractor, list):
+                self.extractor = MultiExtractor(extractor)
```
[Question] What happens when the extractor is a list containing a `MultiExtractor`? Will this line then be:

```python
self.extractor = MultiExtractor([MultiExtractor(...)])
```

Are there any unforeseen consequences?

ninakokalj replied:
I fixed the Pipeline class to be able do take `MultiExtractor` aswell, however if I understand the code correctly the `MultiExctractor class` only takes a `List[ExtractorInterface]` as an attribute and `MultiExtractor` is not an instance of `ExtractorInterface class`, so `MultiExtractor([MultiExtractor(...)])` is not possible.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                >>> pipeline = Pipeline(extractor, strategy)
+
+            Args:
+                extractor (Union[ExtractorInterface, List[ExtractorInterface]]): The extractor to use for entity extraction.
```
In the arguments, there is no need for specifying the argument type, as it is already defined in the function. Please remove all types from the `Args` in all documentations (see other methods in anonipy how it is defined).

Note that the `Attributes` in the class description should have the specified types.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+        """ Anonymize files in the input directory and save the anonymized files to the output directory.
+
+            Args:
+                input_dir (str): The path to the input directory containing files to be anonymized.
```
In the arguments, there is no need for specifying the argument type, as it is already defined in the function. Please remove all types from the Args in all documentations (see other methods in anonipy how it is defined).

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+        """ Anonymize a single file.
+
+            Args:
+                file_path (str): The path to the file to be anonymized.
```
In the arguments, there is no need for specifying the argument type, as it is already defined in the function. Please remove all types from the Args in all documentations (see other methods in anonipy how it is defined).

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+        original_text = open_file(file_path)
+        if original_text is None:
+            print(f"Skipping file {file_path}: Failed to read or file is empty.")
+            return ""
```
It might be worth to discuss/look what to return if the original_text is `None`. Should be actually return an empty string or would it be better to be `None` (i.e. do not generate an anonymized file).

I am more for the `None` option - this way the final list of documents will be the list of only the documents, which were successfully anonymized, without any potential "weird" files.

@ninakokalj What do you think?

ninakokalj replied:
I agree, I’m also in favor of not writing files that were unsuccessfully anonymized. I check in the `anonymize` method if the string I get from `_anonymize_file` (the `anonymized_text` variable) is empty. If it is empty, I skip it and do not write it. But I could do it your way and return `None` if you'd prefer that approach for handling these cases.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+
+        if entities is None:
+            print(f"Skipping file {file_path}: Entity extraction returned None.")
+            return ""
```
[Question] It might be worth to discuss/look what to return if the original_text is `None`. Should be actually return an empty string or would it be better to be `None` (i.e. do not generate an anonymized file).

I am more for the `None` option - this way the final list of documents will be the list of only the documents, which were successfully anonymized, without any potential "weird" files.

@ninakokalj What do you think?

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+
+        original_text = open_file(file_path)
+        if original_text is None:
+            print(f"Skipping file {file_path}: Failed to read or file is empty.")
```
[Future work] Use a different way of printing/logging information about the problems within the package.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+        doc, entities = self.extractor(original_text)
+
+        if entities is None:
+            print(f"Skipping file {file_path}: Entity extraction returned None.")
```
[Future work] Use a different way of printing/logging information about the problems within the package.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+            >>> extractor = NERExtractor(labels, lang=LANGUAGES.ENGLISH)
+            >>> strategy = RedactionStrategy()        
+            >>> pipeline = Pipeline(extractor, strategy)        
+            >>> pipeline.anonymize(r"/path/to/input_dir", r"/path/to/output_dir", False)    
```
Make the example line more descriptive:
```bash
>>> pipeline.anonymize("/path/to/input_dir", "/path/to/output_dir", flatten=False)  
```
Also, do we need the regex expression flag (`r"..."`) to specify the folder paths?

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+        if not os.path.exists(output_dir):
+            os.makedirs(output_dir, exist_ok=True)
+        
+        if os.path.abspath(input_dir) == os.path.abspath(output_dir):
```
Can we move this if statement before the `output_dir` creation commands? So that we do not create the directory before validating if the `input_dir` and `output_dir` folders are the same.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                try:
+                    anonymized_text = self._anonymize_file(file_path)
+                    if not anonymized_text.strip():
+                        print(f"Skipping file {file_path}: Anonymized text is empty.")
```
[Future work] Use a different way of printing/logging information about the problems within the package.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                        print(f"Skipping file {file_path}: Anonymized text is empty.")
+                        continue
+                except Exception as e:
+                    print(f"Error processing file {file_path}: {e}")
```
[Future work] Use a different way of printing/logging information about the problems within the package.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                    continue
+
+                base_name, ext = os.path.splitext(file_name)
+                self.anonymized_files_count += 1
```
Increment the value of `self.anonymized_files_count` after the file is written.

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                    print(f"Error processing file {file_path}: {e}")
+                    continue
+
+                base_name, ext = os.path.splitext(file_name)
```
I think you do not need this `base_name`, because you override it afterwards. If that is true, you can rewrite the line to:

```python
_, ext = os.path.splitext(file_name)
```

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                base_name, ext = os.path.splitext(file_name)
+                self.anonymized_files_count += 1
+                base_name = f"file{self.anonymized_files_count}"
+                output_file_name = f"{base_name}_anony{ext}"
```
You can potentially merge the upper two lines to create the output_file_name:

```python
output_file_name = f"file{self.anonymized_files_count}_anony{ext}"
```
Consider setting the variable `self.anonymized_files_count` as a variable defined in the `anonymize` function (so to have `anonymized_files_count` instead of `self.anonymized_files_count`).

eriknovak commented on anonipy/pipeline.py:
```diff
@@ -0,0 +1,144 @@
+                file_path_after = os.path.relpath(output_file_path, output_dir)
+                file_name_mapping[file_path_before] = os.path.join(output_dir.split(os.sep)[-1], file_path_after)
+
+        return json.dumps(file_name_mapping, indent=4, sort_keys=True)
```
No need to return a string representation of `file_name_mapping`. I would suggest returning it in `dict` format.

eriknovak reviewed (changes requested):
Generally, the code looks good! I added syntactic comments, which should be addressed. The comments labeled as [Question] is something to think about and [Future work] is something to have in mind for the future (as a possible future task).

ninakokalj committed:
Updated pipeline.py with new changes

ninakokalj committed:
Updated pipeline.py with new changes

ninakokalj commented:
I’ve made the requested changes. I’ve added the option to initialize the Pipeline class with the Multiextractor and changed the method `_anonymize_file(self, file_path: str) -> Union[str, None]` to now return `None` if the text is empty or the extraction filed. Please take another look when you have time!

eriknovak reviewed (approved):
Looks good. Merging.
